File: 5-Problem-A-Day/280. Wiggle Sort.java
class Solution {
    public void wiggleSort(int[] nums) {
        boolean less = true;
        for (int i = 0; i < nums.length - 1; i++) {
            if (less) {
                if (nums[i] > nums[i + 1])
                    swap(nums, i, i + 1);
            } else {
                if (nums[i] < nums[i + 1])
                    swap(nums, i, i + 1);
            }
            less = !less;
        }
    }
    
    private void swap(int[] nums, int left, int right) {
        int tmp = nums[left];
        nums[left] = nums[right];
        nums[right] = tmp;
    }
}

File: 2021/280. Wiggle Sort.cpp
class Solution {
public:
    void wiggleSort(vector<int>& nums) {
        int n = nums.size(), right = (n + 1) / 2;
        sort(nums.begin(), nums.end());
        for (int i = 1; i < n && right < n; i += 2) {
            swap(nums[i], nums[right]);
            right++;
        }        
    }
};

class Solution {
public:
    void wiggleSort(vector<int>& nums) {
        int n = nums.size();
        sort(nums.begin(), nums.end());
        for (int i = 1; i < n - 1; i += 2) {
            swap(nums[i], nums[i + 1]);
        }   
    }
};

class Solution {
public:
    void wiggleSort(vector<int>& nums) {
        bool less = true;
        for (int i = 0; i + 1 < nums.size(); i++) {
            if ((less && nums[i] > nums[i + 1]) || (!less && nums[i] < nums[i + 1])) {
                swap(nums[i], nums[i + 1]);
            }
            less = !less;
        }
    }
};

class Solution {
public:
    void wiggleSort(vector<int>& nums) {
        bool smaller = true;
        for (int i = 0; i < (int)nums.size() - 1; i++) {
            if ((smaller && nums[i] > nums[i + 1]) || (!smaller && nums[i] < nums[i + 1])) {
                swap(nums[i], nums[i + 1]);
            }
            smaller = !smaller;
        }
    }
};